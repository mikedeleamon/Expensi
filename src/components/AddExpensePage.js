import React from "react";
import expenses from "../selectors/expenses";
import {connect} from 'react-redux'
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../Actions/expenses";

export class AddExpensePage extends React.Component{
    addExpense=(expense)=>{
        this.props.addExpense(expense)
        {/*history allows auto navigation */}
        this.props.history.push('/')
    }
    render(){
        return(
        <div>
            <h1>Add an Expense</h1>
            <ExpenseForm onSubmit={this.addExpense}/>
        </div>
        )
    }
}


//another way to set up your dispatch calls
const mapDispatchToProps = (dispatch) => {
    return {
        addExpense:(expense) => dispatch(addExpense(expense))
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage)