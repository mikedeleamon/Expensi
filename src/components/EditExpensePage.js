import React from "react";
import {connect} from 'react-redux'
import { editExpense,removeExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class EditExpensePage extends React.Component{
    editExpense =(expense) => {
                this.props.editExpense(this.props.expense.id,expense)
                this.props.history.push('/')
            }
    
    removeExpense = () => {this.props.removeExpense({id:this.props.expense.id})
                            this.props.history.push('/')
                        }

    render(){
        return(
            <div> <h1>Edit an Expense</h1> {this.props.expense.id}
                <ExpenseForm 
                    expense= {this.props.expense}
                    onSubmit = { this.editExpense}

                />
                <button onClick ={this.removeExpense}>remove</button>
            </div>
        
        )
    }  
}

const matchDispacthToProps = (dispatch)=> {
    return{
        // allows you to remove the dispatch commanad by mapping it to props
        editExpense: (id, expense)=> dispatch(editExpense(id,expense)),
        removeExpense: (id) => dispatch(removeExpense(id))

    }
}
    

const mapStateToProps = (state,props) => {
    return{
        expense:state.expenses.find((expense)=> expense.id === props.match.params.id)   
        }
    }

export default connect (mapStateToProps,matchDispacthToProps)(EditExpensePage)