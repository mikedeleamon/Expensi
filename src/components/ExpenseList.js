import React from "react";
import {connect} from 'react-redux'
import ExpsenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses'


export const ExpenseList = (props) => (
    <div>
       <h1> Expense List</h1>
       {props.expenses.length === 0 ? (
        <p>no expense logged</p>
        ):(
        props.expenses.map((expense)=>{
        return( 
        <ExpsenseListItem key={expense.id} {...expense}/>
       )}
       )
       )}
    </div>
)

//set state for connect desired from the store
const mapStatetoProps = (state)=> {
    return {
        expenses:selectExpenses(state.expenses,state.filters)
    }
}

//connect takes in state
//needs provider already set up
export default connect(mapStatetoProps)(ExpenseList)

