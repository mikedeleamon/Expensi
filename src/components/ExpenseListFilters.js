import React from "react";
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import { setFilterText,sortByAmount,sortByDate, setStartDate, setEndDate } from "../actions/filters";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused:null
   }
    
    onDatesChange = ({startDate,endDate}) =>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    
    }

    onFocusChange = (calendarFocused) => (
        this.setState(()=>({calendarFocused}))
    )

    onSelectChange = (e)=> {e.target.value === 'date'? this.props.sortByDate():this.props.sortByAmount()}

    onTextChange = (e)  => {this.props.setFilterText(e.target.value)}

    render(){
    return(
    <div>
        <input type = 'text' placeholder="search" onChange ={this.onTextChange} value={this.props.filters.text}/>
        <select value={this.props.filters.sortBy} onChange = {this.onSelectChange}>
            <option value= 'date' >Date</option>
            <option value= 'amount' >Amount</option>
        </select>

    <DateRangePicker
        startDate={this.props.filters.startDate}
        endDate = {this.props.filters.endDate}
        onDatesChange = {this.onDatesChange}
        focusedInput= {this.state.calendarFocused}
        onFocusChange = {this.onFocusChange}
        numberOfMonths ={1}
        isOutsideRange = {()=>false}
        showClearDates = {true}
    />
    </div>
    )
    }

    


}
const mapStatetoProps = (state) =>(
    {
       filters:state.filters
    }
)

const mapDispatchToProps = (dispatch) => ({
    setFilterText: (text) => dispatch(setFilterText(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: ()=> dispatch(sortByDate())
}
)

export default connect(mapStatetoProps,mapDispatchToProps)(ExpenseListFilters)