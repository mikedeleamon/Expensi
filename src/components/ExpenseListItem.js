import React from 'react'
import moment from 'moment'
import numeral from 'numeral'
import {Link} from 'react-router-dom'


  const ExpsenseListItem = ({id, description, amount,createdAt}) => (
    
    <div>
      <div>
      <Link to={`/edit/${id}`}>
        <h2>{description}</h2>
      </Link>
      </div>
        <h3>{numeral(amount/100).format('$0,0.00')} - created at: {moment(createdAt).format('MMMM Do, YYYY')} </h3>
        
        
    </div>
  )

  export default ExpsenseListItem