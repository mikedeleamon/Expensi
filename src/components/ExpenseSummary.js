import numeral from 'numeral'
import React from "react";
import {connect} from 'react-redux'
import selectExpenses from '../selectors/expenses'
import getExpenseTotal from '../selectors/expenses-total'

 export const ExpsenseSummary = ({expenseCount , expenseTotal}) => {

    const expensePlural = expenseCount === 1 ? 'expense':'expenses'
    const isPlural = expenseCount === 1 ? 'is':'are'
    const formatTotal = numeral(expenseTotal/100).format('$0,0.00')
    return (
    <div>
        <h1>Your total amount of {expenseCount} {expensePlural} {isPlural} {formatTotal}</h1>
    </div>
    )
}

//set state for connect desired from the store
const mapStatetoProps = (state)=> {
    const visibleExpenses = selectExpenses(state.expenses,state.filters)
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: getExpenseTotal(visibleExpenses)
    }
}

// //connect takes in state
// //needs provider already set up
 export default connect(mapStatetoProps)(ExpsenseSummary)
