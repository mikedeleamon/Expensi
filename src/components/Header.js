import React from "react";
import {NavLink} from 'react-router-dom'
import {connect} from 'react-redux'
import { startLogout } from "../actions/auth";


export const Header = ({startLogout}) => (
    <header>
        <h1>Expensi</h1>
        {/* {NavLink allows for styling w/ activeClassName} */}
        <NavLink to='/dashboard' activeClassName='is-active' >Home</NavLink>
        <NavLink to='/create' activeClassName='is-active'>Add Expense</NavLink>
        <NavLink to='/help' activeClassName='is-active'>Help</NavLink>
        <button onClick={startLogout}>Log Out</button>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: ()=> dispatch(startLogout())
})

export default connect(undefined,mapDispatchToProps)(Header)