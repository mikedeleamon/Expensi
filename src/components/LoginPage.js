import React from "react";
import {connect} from 'react-redux'
import { startLogin } from "../actions/auth";

export const LoginPage = ({startLogin}) => (
    <div><button onClick={startLogin}>Log In</button></div>
)

const mapDispatchToprops = (dispatch) => ({
    startLogin: ()=>{dispatch(startLogin())}
})

export default connect(undefined,mapDispatchToprops)(LoginPage)