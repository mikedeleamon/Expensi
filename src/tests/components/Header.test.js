//reactshallowrenderer allows you to test jsx
//Shallow rendering renders parents only
//full rendereing renders parent and child components
//enzyme allows for more complex snapshots and test cases

import React from 'react'
import {Header} from '../../components/Header'
//import ReactShallowRenerer from 'react-test-renderer/shallow'
//import toJSON from 'enzyme-to-json'
import {shallow} from 'enzyme'

let startLogout

beforeEach(()=>{
    startLogout = jest.fn()
    
})

test('Header should render properly',()=>{ 
    //with React renderer
    // const renderer = new ReactShallowRenerer()
    // renderer.render(<Header/>)
    // expect(renderer.getRenderOutput()).toMatchSnapshot()

    //with enzyme
    const wrapper = shallow(<Header startLogout={()=>{}}/>)
    //without serializer config
    //expect(toJSON(wrapper)).toMatchSnapshot()

    //with Serializer config
    expect(wrapper).toMatchSnapshot()
})

test('Header should log out properly',()=>{ 
    
    const wrapper = shallow(<Header  startLogout={startLogout}/>)
    wrapper.find('button').simulate('click')
    //with Serializer config
    expect(startLogout).toHaveBeenCalled()
})

