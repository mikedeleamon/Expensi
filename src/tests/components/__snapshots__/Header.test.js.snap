// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render properly 1`] = `
<header>
  <h1>
    Expensi
  </h1>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    exact={true}
    to="/"
  >
    Home
  </NavLink>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    to="/create"
  >
    Add Expense
  </NavLink>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    to="/help"
  >
    Help
  </NavLink>
  <button
    onClick={[Function]}
  >
    Log Out
  </button>
</header>
`;
